pkg/p2p/libp2p/internal/breaker/breaker.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package breaker
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	_ Interface = (*breaker)(nil)
+
+	// timeNow is used to deterministically mock time.Now() in tests
+	timeNow = time.Now
+
+	// ErrClosed is the special error type that indicates that breaker is closed and that is not executing functions at the moment.
+	ErrClosed = errors.New("breaker closed")
+)
+
+type Interface interface {
+	Execute(f func() error) error
+}
+
+type breaker struct {
+	limit                int
+	consFailedCalls      int
+	firstFailedTimestamp time.Time
+	closedTimestamp      time.Time
+	backoff              time.Duration
+	maxBackoff           time.Duration
+	failInterval         time.Duration
+	mtx                  sync.Mutex
+}
+
+type Options struct {
+	Limit        int
+	FailInterval time.Duration
+	StartBackoff time.Duration
+	MaxBackoff   time.Duration
+}
+
+func NewBreaker(o Options) Interface {
+	breaker := &breaker{
+		limit:        o.Limit,
+		backoff:      o.StartBackoff,
+		maxBackoff:   o.MaxBackoff,
+		failInterval: o.FailInterval,
+	}
+
+	if o.Limit == 0 {
+		breaker.limit = 100
```

Comment:
would be nice to extract these value into `const`s at the head of the source file